libsolidity/experimental/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/experimental/analysis/Analysis.h>
+#include <libsolidity/experimental/analysis/FunctionCallGraph.h>
+
+using namespace solidity::frontend::experimental;
+using namespace solidity::util;
+
+FunctionCallGraph::FunctionCallGraph(solidity::frontend::experimental::Analysis& _analysis):
+	m_analysis(_analysis),
+	m_errorReporter(_analysis.errorReporter()),
+	m_currentNode(nullptr),
+	m_inFunctionDefinition(false)
+{
+}
+
+bool FunctionCallGraph::analyze(SourceUnit const& _sourceUnit)
+{
+	_sourceUnit.accept(*this);
+	// TODO remove debug output before merge
+	//std::cout << annotation().functionCallGraph << std::endl;
+	return !m_errorReporter.hasErrors();
+}
+
+bool FunctionCallGraph::visit(FunctionDefinition const& _functionDefinition)
+{
+	m_inFunctionDefinition = true;
+	m_currentNode = &_functionDefinition;
+	return true;
+}
+
+void FunctionCallGraph::endVisit(FunctionDefinition const&)
+{
+	// If we're done visiting a function declaration without said function referencing/calling
+	// another function in its body - insert it into the graph without child nodes.
+	if (!annotation().functionCallGraph.edges.count(m_currentNode))
+	{
+		annotation().functionCallGraph.edges.insert({m_currentNode, {}});
+		annotation().functionCallGraph.reverseEdges[nullptr].insert(m_currentNode);
```

Comment:
You can just as well insert this into separate members in ``functionCallGraph``  here.
In the end we probably won't need this at all anyways, since we're rather interested in the leaves in the graph of strongly connected components, which are not really in a reasonable relation to these leaf nodes here anyways. So actually, I'd even remove this notion entirely...

Reply:
The notion of the entire reverse graph you mean, not just the set of leaves in the original graph?

Reply:
Whether the reverse graph is helpful or not depends on what you'll need for the SCC algorithm

Reply:
We'll need to walk the SCC forest backwards in the end - but we can also only create the back-edges on that by another forward traversal later on (or use a post-descend-visit in the traversal). So if you don't need the reverse edges for creating a SCC forest, you don't need to keep them here either.

(with my comment I initially just meant encoding the leaves into the ``nullptr`` case, though)

Reply:
Sure. What's our merge strategy on this, i.e. should implemented the SCC forest in this PR, or merge this and then do it in a separate one? If it's the latter, I'd just leave the reverse graph as is, and then get rid of it in the SCC PR if not needed (which it likely won't be).

Reply:
So to be clear: we'll need a directed graph in which the nodes are the strongly connected components of this graph.
That graph will be a forest (so a bunch of trees) that will always have leaves - and in the end we'll need to traverse that forest from the leaves backwards (but for doing so we don't necessarily need to store reverse-edges either, you can visit in a forward-visit and perform your action after descending, effectively giving you a backwards visit using forward edges as well).
For this first graph we need just as much information as is helpful for getting to that visit ultimately.

Reply:
> Sure. What's our merge strategy on this, i.e. should implemented the SCC forest in this PR, or merge this and then do it in a separate one? If it's the latter, I'd just leave the reverse graph as is, and then get rid of it in the SCC PR if not needed (which it likely won't be).

We always want to merge minimal change sets first and as soon as possible.